pilot/pkg/model/validation.go

```diff
@@ -1958,14 +1958,81 @@ func ValidateVirtualService(name, namespace string, msg proto.Message) (errs err
 		}
 	}
 
-	if len(virtualService.Http) == 0 && len(virtualService.Tcp) == 0 {
-		errs = appendErrors(errs, fmt.Errorf("http or tcp must be provided in virtual service"))
+	if len(virtualService.Http) == 0 && len(virtualService.Tcp) == 0 && len(virtualService.Tls) == 0 {
+		errs = appendErrors(errs, errors.New("http, tcp or tls must be provided in virtual service"))
 	}
 	for _, httpRoute := range virtualService.Http {
 		errs = appendErrors(errs, validateHTTPRoute(httpRoute))
 	}
-	// TODO: validate TCP
+	for _, tlsRoute := range virtualService.Tls {
+		errs = appendErrors(errs, validateTLSRoute(tlsRoute))
+	}
+	for _, tcpRoute := range virtualService.Tcp {
+		errs = appendErrors(errs, validateTCPRoute(tcpRoute))
+	}
+
+	return
+}
+
+func validateTLSRoute(tls *networking.TLSRoute) (errs error) {
+	if tls == nil {
+		return nil
+	}
+
+	if len(tls.Match) == 0 {
+		errs = appendErrors(errs, errors.New("TLS route must have at least one match condition"))
+	}
+	for _, match := range tls.Match {
+		errs = appendErrors(errs, validateTLSMatch(match))
+	}
+	if len(tls.Route) != 1 {
+		errs = appendErrors(errs, errors.New("TLS route must have exactly one destination"))
+	}
+	errs = appendErrors(errs, validateDestinationWeights(tls.Route))
```

Comment:
under model.VirtualServices(), we convert short names to FQDNs.. Similarly, convert all TCP blocks into TLS blocks with nil SNI match. This way, the rest of the code is streamlined, as you will only have TLS blocks to deal with.

Reply:
still have to worry about `source_subnet` for TCP, since it doesn't exist in TLS. I feel like the conversion code would end up being almost as long as the code for validating is right now.